File: src/main/java/com/redeprivada/sorting/Quicksort.java
package com.redeprivada.sorting;

public class Quicksort implements Sortable {
    @Override
    public void sort(int[] nums) {
        sort(nums, 0, nums.length - 1);
    }

    private void sort(int[] nums, int begin, int end) {
        if (begin < end) {
            int pivot = partition(nums, begin, end);
            sort(nums, begin, pivot - 1);
            sort(nums, pivot + 1, end);
        }
    }

    private int partition (int[] nums, int begin, int end) {
        int i = begin - 1;
        int pivot = nums[end];
        for (int j = begin; j < end; j++) {
            if (nums[j] < pivot) {
                i++;
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
        }
        i++;
        nums[end] = nums[i];
        nums[i] = pivot;
        return i;
    }
}

File: src/main/kotlin/com/redeprivada/sorting/QuickSort.kt
package com.redeprivada.sorting

class QuickSort : Sortable {
    override fun sort(nums: IntArray) {
        sort(nums, 0, nums.size - 1)
    }

    private fun sort(nums: IntArray, begin: Int, end: Int) {
        if (begin < end) {
            val pivot = partition(nums, begin, end)
            sort(nums, begin, pivot - 1)
            sort(nums, pivot + 1, end)
        }
    }

    private fun partition(nums: IntArray, begin: Int, end: Int): Int {
        var i = begin - 1
        val pivot = nums[end]
        for (j in begin..<end) {
            if (nums[j] < pivot) {
                i++
                val temp = nums[i]
                nums[i] = nums[j]
                nums[j] = temp
            }
        }
        i++
        nums[end] = nums[i]
        nums[i] = pivot
        return i
    }
}
